### ISSUE TITLE
First-class support for project-specific linting rules

### ISSUE THREAD

**Issue by dtinth** [eslint/eslint#8769]
This issue is a continuation of #2715. See that issue for more context.

__tl;dr:__ The current proposal is to `topLevelOptions` property to a config file only available in the project-level config (not shareable/extended configs). Allow rulesDir to be specified there.

## Version

This problem is not solved as of 4.0.0.

## The problem

Many projects have their own unique conventions, [ESLint](https://github.com/eslint/eslint/tree/98512881f1fc2417011247931fa089d987ee8cc6/lib/internal-rules) and [Node](https://github.com/nodejs/node/tree/76340e3f1007998c7cb9d69fa1a42d42663ca6c2/tools/eslint-rules) included. From comments in the previous issue, [Facebook](https://github.com/eslint/eslint/issues/2715#issuecomment-305989282) and [Taskworld](https://github.com/eslint/eslint/issues/2715#issuecomment-300716172) also have project-specific ESLint rules. The `--rulesdir` command-line argument and `rulePaths` APIs can be used to tell ESLint to look for rules in additional directories. Both [Node](https://github.com/nodejs/node/blob/3ad90a4ac8aefad4c5d73fc0607649bf2e2f9cfa/Makefile#L871) and [ESLint](https://github.com/eslint/eslint/blob/4f2f9fcb81446544fb184f0e85dd3912dd2696de/Makefile.js#L61) puts this configuration in their Makefiles.

__Here comes the problem:__ Many people use ESLint integration for their text editors (e.g.
 [Sublime Text](https://github.com/roadhump/SublimeLinter-eslint), [Atom](https://atom.io/packages/linter-eslint)). These plugins don’t read Makefiles; they assumed the project can be linted by running `eslint` in the project directory without passing extra command line arguments or API options.

Even if these editor plugins can be configured to add CLI arguments, they will be applied at global level instead of project-specific level. And these configurations are local to each text editor.

This brings the need for standardized way to include project-specific linting rules that will get picked up by all tools that integrate with ESLint.

### Create an ESLint plugin

This is the approach previously suggested by ESLint maintainers—by creating ESLint plugins and including it in the project. ESLint requires plugins to be in `node_modules` folder. There are several drawbacks to this:

- __Publishing the plugins to `npm`:__ It is cumbersome to update the rules in tandem with the codebase. I would have to go to the plugin, test the rule, and publish a new version to `npm`. The main project needs to point the plugin to the new version. Finally, each developer needs to run `npm install` again to make sure their plugin is at the latest version.
- __Putting the plugin in the project and use `npm link`.__ This makes it cumbersome for developers to prepare the project for development. They would have to run `npm install` inside the plugin folder in addition to the main project folder, and run `npm link` afterwards. If the plugin’s dependencies are updated, they need to run `npm install` inside the plugin folder again.
- __Specifying relative folders in `package.json`.__ This makes it very cumbersome to update plugins, as both `npm@3` and `yarn@0.24` will say “Already up-to-date” when running `yarn` after changing the rules. I need to run `rm -rf node_modules/eslint-plugin-proprietary` before running `yarn` again.

Clearly, using ESLint plugin is more suitable for ESLint rules that are maintained separately. But it is unwieldy for rules that are project-specific.

Here’s how a suitable solution must look like:

- The rules are part of the project. The linting rules are version-controlled along with project’s code and is updated in lockstep.
- They work out-of-the-box with the CLI and all text editor integrations.
- They are ready to use after a single `npm install` command on the project directory.
- They can updated by `git pull` without requiring any extra step.

### Hacking ESLint internals to load extra rules from `.eslintrc.js`

Accroding to @zertosh, [Facebook uses this approach](https://github.com/eslint/eslint/issues/2715#issuecomment-305989282):

```js
// .eslintrc.js
const Rules = require('eslint/lib/rules');
Rules.load(path.join(__dirname, 'resources/eslint-rules'));
```

### Hacking Node’s module resolution system

We at @taskworld put this in `.eslintrc` to hack Node’s module resolution system so that `require('eslint-plugin-proprietary')` becomes `require('./lib/eslint/plugin')`:

```js
// .eslintrc.js
const Module = require('module')

// Hack Node module system to recognize our plugin.
Module._resolveFilename = (original => function (request, requester) {
  if (request === 'eslint-plugin-proprietary') {
    return require.resolve('./lib/eslint/plugin')
  } else {
    return original.apply(this, arguments)
  }
})(Module._resolveFilename)
```

## Proposed solution

[A solution proposed](https://github.com/eslint/eslint/issues/2715#issuecomment-306012754) by @not-an-aardvark is to have `topLevelOptions` property in a configuration file. This `topLevelOptions` is only usable inside a project (cannot be used in shareable/extended configs).

It is a place where we could include the `rulesDir` option, and other options which would not make much sense or would have unintended side-effects when placed in shareable configs.

**Comment by ilyavolodin** [comment]
Would this solve your issue: https://github.com/cletusw/eslint-plugin-local-rules
This is a plugin that allows you to configure local rules through eslintrc file.

**Comment by not-an-aardvark** [comment]
Pinging this issue since I know this was blocking some people from upgrading to ESLint 4.

I'd be interested to know whether `eslint-plugin-local-rules` solves this use case. If so, one potential solution would be to natively support something like that plugin in core (and/or enhance it to avoid needing to use an `./eslint-local-rules.js` file in the project root).

**Comment by dtinth** [comment]
Thanks for pinging.

I will try to evaluate that plugin to see if it solves the problem for us.

**Comment by dtinth** [comment]
@not-an-aardvark I just upgraded ESLint in our project to v4.

The workaround that we use (Hacking Node’s module resolution system) continues to work without problems.

Using `eslint-plugin-local-rules` works in Atom but not in VS Code due to it not `chdir` into ESLint project when invoking ESLint. I opened https://github.com/cletusw/eslint-plugin-local-rules/pull/1 which solves this problem by making the plugin not dependent on `process.cwd()`.

**Comment by not-an-aardvark** [comment]
@dtinth Thanks for looking into that.

I took a shot at implementing a plugin solution to this, with two goals:

* It should be stable (shouldn't rely on modifying ESLint's internals)
* It should only need an ESLint config file, without any additional files or runtime metadata

I ended up with this: [eslint-plugin-rulesdir](https://github.com/not-an-aardvark/eslint-plugin-rulesdir)

The end result is that it works, although it's a bit clumsy (it requires you to use a javascript config file and explicitly load the plugin yourself to set a property on it). Still, it's probably better than hacking Node's module resolution or ESLint's plugin resolution.

The main challenge is that the plugin needs to know user's rules directory in order to export a `rules` object, but I don't know of a way for a config file to provide that metadata declaratively, since the plugin doesn't have a reference to the config file. If we do enhance ESLint core to support this, one way to do it would be to allow that metadata to be provided somehow.

In theory, if an organization had a consistent project structure with all the ESLint rules at the same local path for each project, they could also fork the plugin and hardcode the path, which would make it "just work" without needing to do extra things in the config file.

Another option would be for the plugin to look for a property in `package.json` in the CWD, but that would prevent eslint from being run with a different CWD (which would cause the same issue with the VSCode integration).

**Comment by dtinth** [comment]
@not-an-aardvark I ended up creating [`eslint-plugin-local`](https://www.npmjs.com/package/eslint-plugin-local), heavily inspired by `eslint-plugin-local-rules`. The module is one line of code:

```js
module.exports = require('../../.eslintplugin')
```

Since this plugin is intended to be installed at project-level (not in shareable config; in that case, better to create a plugin), we can be sure it will reside at `$PROJECT/node_modules/eslint-plugin-local`. This means the plugin will work regardless of the `cwd`.

The name `.eslintplugin` is chosen so that the file appears alongside `.eslintrc`. Then we can implement our project-specific plugin at:

- `.eslintplugin.js`
- `.eslintplugin/index.js`

**Comment by not-an-aardvark** [comment]
Neat, that looks useful (and probably better than my version for projects that want to have full plugin functionality rather than just simulating `--rulesdir`).

Given that the problem seems to be solvable with plugins, my impression is that no core change is needed. Do you agree? (Also asking @zertosh since he was trying to make this work for Facebook in https://github.com/eslint/eslint/issues/2715#issuecomment-305989282.)

**Comment by dtinth** [comment]
This solved the problem for me.

Thanks @not-an-aardvark and @cletusw for directing me to a simple solution. 😄

**Comment by zertosh** [comment]
Sorry this slipped through my queue. I did a quick review internally and 3 things stood out:

1. Switching from `--rulesdir` to the plugin hack is going to require codemoding all of the ignore comments to add `rulesdir/` to the name. Not great, but it's a one-time cost, so I guess I can live with that.
2. This solves the immediate problem (which is great! thank you!), but I found lots of cases that would still benefit from being able to configure other CLI flags via a project-level config:
    * Setting `--extensions`. Although most js files are `.js`, others are`.jsx`, but a lot are `.js.flow` (flow lib defs that live next to their implementation - this is useful for files that can't be run through a preprocessor to strip types).
    * Setting `--max-warnings=0`.
3. Some projects don't have any `node_modules` at all, but rather use a globally available eslint. They don't use any custom plugins, but do have custom rules. So they don't have a way to make use of `eslint-plugin-rulesdir`. I get that this a weird use case, so I wouldn't weigh this too heavily.

I'm still very open to implementing this proposal, if it's accepted.

**Comment by not-an-aardvark** [comment]
Thanks for the feedback.

* Regarding `--extensions`: In a [TSC meeting](https://github.com/eslint/tsc-meetings/blob/0b69c20b635e569c405c03b0ea63dff8026c00f8/notes/2017/2017-05-11.md#breaking-add-jsx-to-the-default---ext-file-extensions-fixes-8399-8515) a few months ago, there was some agreement that it would be a good idea to allow plugins to declare specific extensions that they support (pending a proposal for the specific API). If that gets added, the `--extensions` flag might end up being mostly obsolete.
* Regarding `--max-warnings=0`, I'm confused about your use case. Isn't it already possible to do this in a config file by simply configuring all the rules to errors rather than warnings?
* Regarding global installations with `eslint-plugin-rulesdir`: would that be solved if you modified the global `eslint-plugin-rulesdir` in your config file instead?

    ```diff
    -const rulesDirPlugin = require('eslint-plugin-rulesdir');
    +const rulesDirPlugin = module.parent.require('eslint-plugin-rulesdir');
    rulesDirPlugin.RULES_DIR = 'tools/eslint-rules';
    ```

    One other way to resolve that would be to change the `eslint-plugin-rulesdir` API to make the config file set a property on the global object. (This usually isn't a good idea, but if a config needs to modify global state anyway by setting a property on another module, it might be easier to just use the global object directly.)

**Comment by zertosh** [comment]
>  If that gets added, the --extensions flag might end up being mostly obsolete.

Good to know! 👍 

> Regarding --max-warnings=0, I'm confused about your use case. Isn't it already possible to do this in a config file by simply configuring all the rules to errors rather than warnings?

We like errors for syntax issues (something that will prevent your code from running), and warnings for style/correctness issues. But neither should be committed.

> Regarding global installations with eslint-plugin-rulesdir: would that be solved if you modified the global eslint-plugin-rulesdir in your config file instead?

Yup, that'll work! I'll switch [nuclide](https://github.com/facebook/nuclide/blob/693623fd6e2124b094ab4d4d5fad15a67496c6b7/.eslintrc.js#L513-L514) this week to `eslint-plugin-rulesdir`. If that goes well, then I'll roll it out to RN (which is the one I'm worried about because of the # of ignores to rewrite). 

Thank for this @not-an-aardvark!

**Comment by not-an-aardvark** [comment]
Great, thanks. In that case I'll close this issue.
